### DISCUSSION TITLE
GCNConv and parallel execution on a CPU

### DISCUSSION THREAD

**Discussion by l-hoang** [pyg-team/pytorch_geometric#3900] [Q&A]
Hello.

I've been experimenting with PyTorch Geometric's GCNConv layer on CPUs, and it seems like
they do not leverage much parallelism when given multiple threads.

Here is the code I use:
```
class GCN(torch.nn.Module):
    def __init__(self, in_dim, out_dim):
        super().__init__()
        self.hidden_feat = 2000

        self.conv1 = GCNConv(in_dim, self.hidden_feat)
        self.conv2 = GCNConv(self.hidden_feat, self.hidden_feat)
        self.conv3 = GCNConv(self.hidden_feat, out_dim)

    def forward(self, data):
        x, edge_index = data.x, data.edge_index

        x = self.conv1(x, edge_index)
        x = self.conv2(x, edge_index)
        x = self.conv3(x, edge_index)

        return x.log_softmax(dim=1)
```
3 GCN layers back to back with a hidden feature size of 2000.

If I run this with 1 thread (confirmed with torch.get_num_threads()) the forward phases takes 68 seconds.
The data batch print shown is the number of rows and edges in the batch for reference.

```
Data(x=[96185, 128], edge_index=[2, 255628], y=[96185], seed_nodes=[10000])
model run done 68.32124781608582
```

If I run with 32 threads (the number of cores on the machine I am running on), the time improves, but
not by much (note the data batch is different since sampling occurs randomly, but it's in the same ballpark):

```
Data(x=[96091, 128], edge_index=[2, 254745], y=[96091], seed_nodes=[10000])
model run done 49.60090661048889
```

How much parallelism is expected from the GCNConv layer? It seems like going from 1 thread to 32 threads doesn't
improve the runtime much, and from my observations with `top`, most of the time seems to be spent
not using all the allocated threads (i.e. there doesn't seem to be much parallel execution going on).

Thank you.

**Comment by rusty1s** [comment]
As far as I know, `torch.scatter_add` does not have great multi-threading support on CPU, which probably explains your observed runtimes. That is, you only get minor benefits in efficiency on CPU through other ops (matmul, gather, ...), while  most of the time is spent in single-threaded scatter ops. It might be a good idea to do a micro benchmark of `torch.scatter_(..., reduce="sum")` and `torch_scatter.scatter` and submit it to the PyTorch and `torch-scatter` repos.

**Reply by bowu** [reply]
> As far as I know, `torch.scatter_add` does not have great multi-threading support on CPU, which probably explains your observed runtimes. That is, you only get minor benefits in efficiency on CPU through other ops (matmul, gather, ...), while most of the time is spent in single-threaded scatter ops. It might be a good idea to do a micro benchmark of `torch.scatter_(..., reduce="sum")` and `torch_scatter.scatter` and submit it to the PyTorch and `torch-scatter` repos.

What's the reason to use torch-scatter when torch already has parallel scatter operations?

**Reply by rusty1s** [reply]
PyTorch does currently only provide scatter support for reductions `"sum"` and `"mul"`. This library adds support for a broader range of reductions, as well as support for various `segment_*` operations. There exists some plans to fully incorporate this library into PyTorch, see https://github.com/pytorch/pytorch/issues/63780.

**Reply by bowu** [reply]
@rusty1s Is there a plan to parallelize `pytorch_scatter` for CPUs?

**Reply by rusty1s** [reply]
Yes, definitely. Either as part of the PyTorch merge or directly inside `torch-scatter`.

**Reply by bowu** [reply]
@rusty1s Is it possible to let us know an estimate on the latest time to have parallel scatter operations to back up PyG? CPU training performance is important to us. Thanks!

**Reply by rusty1s** [reply]
I don't have any estimate on this one, I'm sorry. How about we create an issue in `torch-scatter` to track progress on this one? Let me know if you are interested in helping us integrating this feature.

I think this should be pretty straightforward to add via `at::parallel_for`, but we need to make sure that we do not override results from different threads that want to write to the same location.

**Reply by rusty1s** [reply]
It's also mentioned here: https://github.com/pyg-team/pytorch_geometric/issues/2306. We will track any progress in this issue.

**Reply by bowu** [reply]
> It's also mentioned here: https://github.com/pyg-team/pytorch_geometric/issues/2306. We will track any progress in this issue.

Is the temporary solution to use SparseTensor? Its scatter operation is parallel. Any caveat we should know?

**Reply by rusty1s** [reply]
Not all GNN layers can utilize sparse-matrix multiplication within `SparseTensor`. As such, the current temporary solution is to use GPUs :)

Me and @mananshah99 will try to integrate that in a new release once PyTorch 1.11 lands.

**Reply by bowu** [reply]
@rusty1s Due to some reason, we need to focus on CPUs now. Could you let us know why the backward function is not parallelized but the forward function is as shown in this [file](https://github.com/rusty1s/pytorch_sparse/blob/3e87af1cc2218954fd6fd457f96ab4ef6e0f3b15/csrc/cpu/spmm_cpu.cpp#L103)? If we'd like to leverage `torch_sparse` and optimized CPU performance, what should we do? Do we convert the edge tensor in the COO format to a SparseTensor? Is there anything extra we should do to maximize performance?

**Reply by rusty1s** [reply]
Parallel `scatter` support should be available once https://github.com/rusty1s/pytorch_scatter/pull/270 lands (thanks to @mananshah99).

The `spmm_value_bw_cpu` computes the gradients w.r.t. to the edge weights which is only used in case of attention. The gradients w.r.t. to node features use `spmm` under-the-hood and should make use of parallelization already.

**Reply by bowu** [reply]
> Parallel `scatter` support should be available once https://github.com/rusty1s/pytorch_scatter/pull/270 lands (thanks to @mananshah99).
> 
> 
> 
> The `spmm_value_bw_cpu` computes the gradients w.r.t. to the edge weights which is only used in case of attention. The gradients w.r.t. to node features use `spmm` under-the-hood and should make use of parallelization already.

After it lands, should we use COO edge tensor or SparseTensor for the best performance?

**Reply by rusty1s** [reply]
This is hard to tell and depends on the use-case. For dense graphs, it is definitely advisable to use `SparseTensor`.

**Reply by bowu** [reply]
I thought `SparseTensor` is optimized for sparse graphs. Could you let us know why it mainly works for dense graphs?

**Reply by rusty1s** [reply]
That's more due to reasons of memory consumption. If you have a dense graph, then the usage of`SparseTensor` may be able to avoid materializing edge-level message representations (dependent on the GNN operator). See [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html) for more details.
